File: shared/src/utils/pricing.ts
/**
 * Regras de precificação compartilhadas entre backend (Agente 5) e frontend
 * (pré-visualização). Puras e testáveis.
 */

export interface MarkupTiers {
  tier1: number; // até 30
  tier2: number; // 30–100
  tier3: number; // 100–300
  tier4: number; // acima de 300
}

export const DEFAULT_MARKUP: MarkupTiers = {
  tier1: 1.2,
  tier2: 0.9,
  tier3: 0.7,
  tier4: 0.5,
};

/** Seleciona o markup pela faixa do preço de compra. */
export function selectMarkup(purchasePrice: number, tiers: MarkupTiers = DEFAULT_MARKUP): number {
  if (purchasePrice <= 30) return tiers.tier1;
  if (purchasePrice <= 100) return tiers.tier2;
  if (purchasePrice <= 300) return tiers.tier3;
  return tiers.tier4;
}

/**
 * Arredonda para "preço psicológico" terminando em ,90.
 * Ex.: 72,40 → 79,90 | 118,00 → 119,90 | 210,10 → 219,90
 * Nunca retorna número quebrado.
 */
export function toPsychologicalPrice(value: number): number {
  if (value <= 0) return 0;
  // sobe para o próximo "inteiro de dezena base" e aplica ,90
  const whole = Math.ceil(value);
  const endsBelow90 = whole - Math.floor(whole / 10) * 10 <= 9;
  const base = endsBelow90 ? Math.floor(whole / 10) * 10 + 9 : Math.ceil(whole / 10) * 10 + 9;
  const candidate = base + 0.9;
  return candidate < value ? candidate + 10 : Number(candidate.toFixed(2));
}

export interface ComputePriceInput {
  purchasePrice: number;
  marketAveragePrice?: number;
  tiers?: MarkupTiers;
}

export interface ComputedPrice {
  suggestedPrice: number;
  markupApplied: number;
  profit: number;
  marginPercent: number;
  roi: number;
}

/**
 * Calcula preço sugerido, lucro, margem e ROI.
 * Se houver preço médio de mercado, o preço-alvo é limitado para não destoar
 * (não passa de 10% acima da média), mantendo competitividade.
 */
export function computePrice(input: ComputePriceInput): ComputedPrice {
  const { purchasePrice, marketAveragePrice, tiers } = input;
  const markup = selectMarkup(purchasePrice, tiers);
  let target = purchasePrice * (1 + markup);

  if (marketAveragePrice && marketAveragePrice > 0) {
    const ceiling = marketAveragePrice * 1.1;
    target = Math.min(target, ceiling);
    target = Math.max(target, purchasePrice * 1.05); // nunca abaixo de 5% de lucro
  }

  const suggestedPrice = toPsychologicalPrice(target);
  const profit = Number((suggestedPrice - purchasePrice).toFixed(2));
  const marginPercent = Number(((profit / suggestedPrice) * 100).toFixed(2));
  const roi = Number(((profit / purchasePrice) * 100).toFixed(2));

  return { suggestedPrice, markupApplied: markup, profit, marginPercent, roi };
}

File: backend/app/shared/utilities/pricing.py
from pydantic import BaseModel


class MarkupTiers(BaseModel):
    tier1: float
    tier2: float
    tier3: float
    tier4: float


DEFAULT_MARKUP = MarkupTiers(tier1=1.2, tier2=0.9, tier3=0.7, tier4=0.5)


def select_markup(purchase_price: float, tiers: MarkupTiers = DEFAULT_MARKUP) -> float:
    if purchase_price <= 30:
        return tiers.tier1
    if purchase_price <= 100:
        return tiers.tier2
    if purchase_price <= 300:
        return tiers.tier3
    return tiers.tier4


def to_psychological_price(value: float) -> float:
    if value <= 0:
        return 0

    import math

    whole = math.ceil(value)
    base = math.floor(whole / 10) * 10 + 9
    candidate = base + 0.9
    if candidate < value:
        candidate += 10
    return round(candidate, 2)


class ComputePriceInput(BaseModel):
    purchasePrice: float
    marketAveragePrice: float | None = None
    tiers: MarkupTiers | None = None


class ComputedPrice(BaseModel):
    suggestedPrice: float
    markupApplied: float
    profit: float
    marginPercent: float
    roi: float


def compute_price(input_: ComputePriceInput) -> ComputedPrice:
    purchase_price = input_.purchasePrice
    markup = select_markup(purchase_price, input_.tiers or DEFAULT_MARKUP)
    target = purchase_price * (1 + markup)

    if input_.marketAveragePrice and input_.marketAveragePrice > 0:
        ceiling = input_.marketAveragePrice * 1.1
        target = min(target, ceiling)
        target = max(target, purchase_price * 1.05)

    suggested_price = to_psychological_price(target)
    profit = round(suggested_price - purchase_price, 2)
    margin_percent = round((profit / suggested_price) * 100, 2) if suggested_price else 0
    roi = round((profit / purchase_price) * 100, 2) if purchase_price else 0

    return ComputedPrice(
        suggestedPrice=suggested_price,
        markupApplied=markup,
        profit=profit,
        marginPercent=margin_percent,
        roi=roi,
    )
